File: Problem_1 Rotten_Oranges.py
# // Time Complexity : O(m*n)
# // Space Complexity : O(m*n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
#
#
# // Your code here along with comments explaining your approach

class Solution:
    def orangesRotting(self, grid):
        if grid == None or len(grid) == 0:
            return 0
        fresh = 0
        time = 0
        m = 0
        n = 0
        m = len(grid)
        n = len(grid[0])
        q = deque()

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:     # Append Rotten oranges to the queue
                    q.append((i, j))
                if grid[i][j] == 1:
                    fresh += 1          #Calculate Total fresh ones

        dirs = [(0, 1), (0, -1), (-1, 0), (1, 0)]  # 4 direction array
        if fresh == 0:
            return 0
        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                for d in dirs:
                    r = d[0] + curr[0]      #Considering 4 arrays
                    c = d[1] + curr[1]
                    #rot if fresh,
                    #Append rotten to queue
                    if r >= 0 and r < m and c >= 0 and c < n and grid[r][c] == 1:
                        grid[r][c] = 2
                        q.append((r, c))
                        fresh -= 1  #Decrement Fresh ones
            time += 1
        if fresh != 0:
            return -1
        return time - 1

File: Problem1_RottenOranges.java

// Time Complexity : o(mn)
// Space Complexity : o(mn)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int orangesRotting(int[][] grid) {
        
        if(grid==null || grid.length==0)
            return 0;
        
        int fresh=0;
        int time=0;
        int m=grid.length;
        int n=grid[0].length;
        
        Queue<Integer> queue=new LinkedList<>();
        
        //find rotten oranges and fresh oranges
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(grid[i][j]==2)
                {
                    queue.add(i);
                    queue.add(j);
                }
                else if(grid[i][j]==1)
                    fresh++;
            }
        }
        
        if(fresh==0)
            return 0;
        
        int dirs[][]={{-1,0},{0,-1},{0,1},{1,0}};
        
        while(!queue.isEmpty())
        {
            int size=queue.size();
            
            for(int k=0;k<size;k+=2)
            {
                int r=queue.poll();
                int c=queue.poll();
                
                for(int[] dir:dirs)
                {
                    int nR=r+dir[0];
                    int nC=c+dir[1];
                    
                    if(nR>=0 && nR<m && nC>=0 && nC<n && grid[nR][nC]==1)
                    {
                       grid[nR][nC]=2; 
                        fresh--;
                        queue.add(nR);
                        queue.add(nC);   
                    }
                           
                }
            }
            time++;
        }
        if(fresh!=0)
            return -1;
        
        return time-1;
    }
}